### PR TITLE
add define to indicate that gcc is nonstandard for llvm, keep .as in addition to .sos

### PR DESCRIPTION
**Pull request by davidlange6**
These proposed changes are to move towards a more standardized llvm build.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by smuzaffar** [comment]
test parameters:
- full_cmssw = true

**Comment by smuzaffar** [comment]
please test for CMSSW_12_6_CLANG_X

File: llvm.spec [hunk 1]
```diff
@@ -80,8 +81,6 @@ rm -f %{i}/bin/FileRadar.scpt %{i}/bin/GetRadarVersion.scpt
 
 # Avoid dependency on /usr/bin/python, Darwin + Xcode specific
 rm -f %{i}/bin/set-xcode-analyzer
-# Remove all static libs except libomptarget-nvptx.a
-find %{i}/lib64 -name "*.a" -and -not -name "libomptarget-*.a" -delete
```
Comment (smuzaffar) [inline llvm.spec:84]: @davidlange6 , note that this adds additional 900MB to distribution

**Comment by smuzaffar** [comment]
+externals

File: llvm.spec [hunk 2]
```diff
@@ -38,6 +38,7 @@ cd %{_builddir}/build
 
 cmake %{_builddir}/llvm-%{realversion}-%{llvmCommit}/llvm \
   -G Ninja \
+  -DGCC_INSTALL_PREFIX="${GCC_ROOT}" \
```
Comment (smuzaffar) [inline llvm.spec:41]: @fwyzard , looks like building llvm with `GCC_INSTALL_PREFIX`  fixed the llvm not finding our gcc header issue. After this change now I see that `clang++` shows the correct gcc paths
```
>  clang++ -E -x c++ - -v
clang version 14.0.6
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/bin
Found candidate GCC installation: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0
Selected GCC installation: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /cvmfs/cms-ib.cern.ch/nweek-02748/el8_amd64_gcc10/external/cuda/11.5.2-c927b7e765e06433950d8a7eab9eddb4, version 11.5
 (in-process)
 "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/bin/clang-14" -cc1 -triple x86_64-redhat-linux-gnu -E -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -v -fcoverage-compilation-dir=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_6_CLANG_X_2022-08-31-2300 -resource-dir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6 -internal-isystem /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0 -internal-isystem /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0/x86_64-redhat-linux-gnu -internal-isystem /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0/backward -internal-isystem /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6/include -internal-isystem /usr/local/include -internal-isystem /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../x86_64-redhat-linux-gnu/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_6_CLANG_X_2022-08-31-2300 -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o - -x c++ -
clang -cc1 version 14.0.6 based upon LLVM 14.0.6 default target x86_64-redhat-linux-gnu
ignoring nonexistent directory "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../x86_64-redhat-linux-gnu/include"
ignoring nonexistent directory "/include"
#include "..." search starts here:
#include <...> search starts here:
 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0
 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0/x86_64-redhat-linux-gnu
 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/lib/gcc/x86_64-redhat-linux-gnu/10.3.0/../../../../include/c++/10.3.0/backward
 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6/include
 /usr/local/include
 /usr/include
End of search list.
```

Comment (smuzaffar) [reply llvm.spec:41]: I still need to see if it works properly when it is installed on cvmfs

**Comment by davidlange6** [comment]
> 
> 
> @fwyzard , looks like building llvm with GCC_INSTALL_PREFIX fixed the llvm not finding our gcc header issue. After this change now I see that clang++ shows the correct gcc paths

Yes, that is indeed what this flag fixes...

**Comment by smuzaffar** [comment]
humm, not really. @davidlange6 it works in the build area ( where gcc install path and GCC_INSTALL_PREFIX are same) but once things are installed on cvmfs then clang does not find our gcc ( specially in llvm 14)
```
Singularity> which clang++
/cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/bin/clang++

Singularity> which gcc
/cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/gcc/10.3.0-84898dea653199466402e67d73657f10/bin/gcc

Singularity> echo '' | clang++ -E -x c++ - -v
clang version 14.0.6
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/bin
Found CUDA installation: /cvmfs/cms-ib.cern.ch/nweek-02748/el8_amd64_gcc10/external/cuda/11.5.2-c927b7e765e06433950d8a7eab9eddb4, version 11.5
 (in-process)
 "/cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/bin/clang-14" -cc1 -triple x86_64-redhat-linux-gnu -E -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -v -fcoverage-compilation-dir=/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmsdist/7729/27252/CMSSW_12_6_CLANG_X_2022-08-31-2300 -resource-dir /cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6 -internal-isystem /cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6/include -internal-isystem /usr/local/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir=/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmsdist/7729/27252/CMSSW_12_6_CLANG_X_2022-08-31-2300 -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o - -x c++ -
clang -cc1 version 14.0.6 based upon LLVM 14.0.6 default target x86_64-redhat-linux-gnu
ignoring nonexistent directory "/include"
#include "..." search starts here:
#include <...> search starts here:
 /cvmfs/cms-ci.cern.ch/week0/PR_334f39cf/el8_amd64_gcc10/external/llvm/14.0.6-7e877bac6aa04ac511ab1c515ec417ab/lib64/clang/14.0.6/include
 /usr/local/include
 /usr/include
End of search list.
```

**Comment by davidlange6** [comment]
Eg, it doesn't work once gcc is relocated.. (this I didn't test). Hum

**Comment by smuzaffar** [comment]
looks like llvm 13 and above ignore the `COMPILER_PATH` env see https://github.com/llvm/llvm-project/issues/51351

**Comment by smuzaffar** [comment]
we can either patch llvm and add back `SmallVector<std::string, 8> Prefixes(D.PrefixDirs.begin(),D.PrefixDirs.end());` or  start using `--gcc-toolchain=$GCC_ROOT`

**Comment by fwyzard** [comment]
For the external version of llvm/clang coming with HIP, I'm passing `--gcc-toolchain=$COMPILER_PATH` explicitly.

**Pull request merged**